src/go/kreq-gen/kafka-request-generator.go

```diff
@@ -0,0 +1,239 @@
+	records := makeRandomRecords()
+	var bytearr []byte
+	for _, record := range records {
+		bytearr = append(bytearr, record.AppendTo(nil)...)
+	}
+	recordBatch.Records = bytearr
+	recordBatch.NumRecords = int32(len(records))
+	recordBatch.Magic = int8(2)       // new format
+	recordBatch.Attributes = int16(0) // uncompressed
+	rawBatch := recordBatch.AppendTo(nil)
+	recordBatch.Length = int32(len(rawBatch[8+4:]))
+	// skip first offset (int64) and length
+	recordBatch.CRC = int32(crc32.Checksum(rawBatch[8+4+4+1+4:], crc32c))
+	return recordBatch
+}
+
+func makeRandomKafkaData(version int16) []byte {
+	var bytearr []byte
+	if version >= 3 {
+		data := makeRandomRecordBatches()
+		bytearr = data.AppendTo(nil)
+	} else if version == 2 {
+		data := makeRandomLegacyBatchV1()
+		for _, record := range data {
+			bytearr = append(bytearr, record.AppendTo(nil)...)
+		}
+	} else {
+		data := makeRandomLegacyBatchV0()
+		for _, record := range data {
+			bytearr = append(bytearr, record.AppendTo(nil)...)
+		}
+	}
+	return bytearr
+}
+
+func randomFill(v reflect.Value, version int16) {
+	switch v.Kind() {
+	case reflect.Ptr:
+		v.Set(reflect.New(v.Type().Elem()))
+		/// This conditional represents setting an nullable field to nil
+		/// happening 50% of the time
+		if makeRandomBool() {
+			randomFill(v.Elem(), version)
+		}
+	case reflect.Struct:
+		for i := 0; i < v.NumField(); i++ {
+			randomFill(v.Field(i), version)
+		}
+		// Special cases, the following byte arrays must be populated with kafka
+		// record & recordBatch formatted data
+		if v.Type().Name() == "ProduceRequestTopicPartition" {
+			recordBatchData := makeRandomKafkaData(version)
+			v.FieldByName("Records").Set(reflect.ValueOf(recordBatchData))
+		} else if v.Type().Name() == "FetchResponseTopicPartition" {
+			recordBatchData := makeRandomKafkaData(version)
+			v.FieldByName("RecordBatches").Set(reflect.ValueOf(recordBatchData))
+		}
+	case reflect.Array:
+		num := v.Len()
+		arr := reflect.New(reflect.ArrayOf(num, v.Type().Elem())).Elem()
+		v.Set(arr)
+		for i := 0; i < num; i++ {
+			randomFill(v.Index(i), version)
+		}
+	case reflect.Slice:
+		num := rand.Intn(10)
+		v.Set(reflect.MakeSlice(v.Type(), num, num))
+		for i := 0; i < num; i++ {
+			randomFill(v.Index(i), version)
+		}
+	case reflect.Bool:
+		v.SetBool(makeRandomBool())
+	case reflect.Uint8:
+		v.SetUint(uint64(makeRandomInt64()))
+	case reflect.Int8:
+		fallthrough
+	case reflect.Int16:
+		fallthrough
+	case reflect.Int32:
+		fallthrough
+	case reflect.Int64:
+		v.SetInt(makeRandomInt64())
+	case reflect.String:
+		v.SetString(makeRandomString())
+	case reflect.Map:
+		/// Only case where Map is an acceptable type
+		tags := makeRandomTags()
+		/// The tags structure is an unexported private field
+		/// Use `unsafe.Pointer` to be able to set it
+		rf := reflect.NewAt(v.Type(), unsafe.Pointer(v.UnsafeAddr())).Elem()
+		rf.Set(reflect.MakeMapWithSize(reflect.TypeOf(map[uint32][]byte{}), tags.Len()))
```

Comment:
The unexported aspect in kmsg is deliberate -- I may change the guts within the future. I don't think you need to do this, though. You could change the reflect.Struct case above to check if the type is Tags, and then do this logic up there.